File: Week_04/find_largest_value_in_each_tree_row/Solution.java
package find_largest_value_in_each_tree_row;

import binary_tree.TreeNode;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;

/**
 * DFS
 */
class Solution1 {
    public List<Integer> largestValues(TreeNode root) {
        List<Integer> max = new ArrayList<>();
        dfs(root, 0, max);
        return max;
    }

    private void dfs(TreeNode root, int level, List<Integer> max) {
        if (root == null) return;

        if (max.size() == level) max.add(root.val);
        else max.set(level, Math.max(root.val, max.get(level)));

        dfs(root.left, level + 1, max);
        dfs(root.right, level + 1, max);
    }
}

/**
 * BFS
 */
class Solution2 {
    public List<Integer> largestValues(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        if (root == null) return res;

        Queue<TreeNode> q = new ArrayDeque<>();
        q.add(root);
        while (!q.isEmpty()) {
            int n = q.size();
            int max = Integer.MIN_VALUE;
            for (int i = 0; i < n; i++) {
                TreeNode node = q.poll();
                if (node.val > max) max = node.val;
                if (node.left != null) q.offer(node.left);
                if (node.right != null) q.offer(node.right);
            }
            res.add(max);
        }
        return res;
    }
}

File: Week_04/find_largest_value_in_each_tree_row/Solution.py
# DFS
class Solution1:
    def largestValues(self, root: TreeNode) -> List[int]:
        def dfs_compare(root, level = 0):
            if root is None:
                return
            if len(max_list) == level:
                max_list.append(-sys.maxsize)
            if root.val > max_list[level]:
                max_list[level] = root.val
            dfs_compare(root.left, level + 1)
            dfs_compare(root.right, level + 1)

        max_list = []
        dfs_compare(root)
        return max_list


# BFS
class Solution2:
    def largestValues(self, root: TreeNode) -> List[int]:
        max_list = []
        dq = collections.deque([root]) if root else []
        while dq:
            max_val = -sys.maxsize
            size = len(dq)
            for _ in range(size):
                node = dq.popleft()
                max_val = max(max_val, node.val)
                if node.left: dq.append(node.left)
                if node.right: dq.append(node.right)
            max_list.append(max_val)
        return max_list


# more concise and pythonic code
class Solution2:
    def largestValues(self, root: TreeNode) -> List[int]:
        max_list = []
        row = [root]
        while any(row):
            max_list.append(max(node.val for node in row))
            row = [child for node in row for child in (node.left, node.right) if child]
        return max_list
